### PR TITLE
API initialization changes

### LINKED ISSUE #482: getsockopt: connection refused
Getting `ERRO[0000] Could not fetch task                          error="Get http://127.0.0.1:8080/tasks: dial tcp 127.0.0.1:8080: getsockopt: connection refused" runner=async` when starting container

### PR DESCRIPTION
**Pull request by pedronasser**
- Fixed env PORT was not working correctly because viper.AutomaticEnv was positioned in the wrong place.
- API serve not using gin.Engine.Run because using that method we can't know when the server is already listening in the specified port.
- Some log changes

Closes #482

### CODE REVIEW

File: api/server/server.go [hunk 1]
```diff
@@ -18,6 +19,8 @@ import (
 	"github.com/iron-io/functions/api/runner/task"
 	"github.com/iron-io/functions/api/server/internal/routecache"
 	"github.com/iron-io/runner/common"
+	"github.com/spf13/viper"
+	"net"
```
Comment (ccirello) [inline api/server/server.go]: Please organize this import blog.

File: api/server/server.go [hunk 2]
```diff
@@ -187,18 +190,18 @@ func (s *Server) startGears(ctx context.Context) {
...
-		go s.Router.Run()
+		listen := fmt.Sprintf(":%d", viper.GetInt(EnvPort))
+		listener, err := net.Listen("tcp", listen)
+		if err != nil {
+			logrus.WithError(err).Fatalln("Failed to serve functions API.")
+		}
+		logrus.Infof("Serving Functions API on address `%s`", listen)
+		go runner.StartWorkers(ctx, s.Runner, s.tasks)
+		go runner.RunAsyncRunner(ctx, s.apiURL, s.tasks, s.Runner)
+		go http.Serve(listener, s.Router)
```
Comment (ccirello) [inline api/server/server.go]: This is all wrong. srv.AddFunc should be managing individual services. Here it is managing all of them together. 👎

File: api/server/server.go [hunk 3]
```diff
@@ -177,26 +180,35 @@ func (s *Server) Start(ctx context.Context) {
...
-	// PORT environment variable was defined.
 	svr.AddFunc(func(ctx context.Context) {
-		go s.Router.Run()
+		go http.Serve(listener, s.Router)
 		<-ctx.Done()
 	})
 
 	svr.AddFunc(func(ctx context.Context) {
-		runner.StartWorkers(ctx, s.Runner, s.tasks)
+		go runner.RunAsyncRunner(ctx, s.apiURL, s.tasks, s.Runner)
```
Comment (ccirello) [inline api/server/server.go]: if I understand correctly, `runner.RunAsyncRunner` is blocking already. No need for `go` plus reading the `ctx.Done()` channel.

File: api/server/server.go [hunk 4]
```diff
@@ -177,26 +180,35 @@ func (s *Server) Start(ctx context.Context) {
...
+		<-ctx.Done()
 	})
 
 	svr.AddFunc(func(ctx context.Context) {
-		runner.RunAsyncRunner(ctx, s.apiURL, s.tasks, s.Runner)
+		go runner.StartWorkers(ctx, s.Runner, s.tasks)
+		<-ctx.Done()
```
Comment (ccirello) [inline api/server/server.go]: ditto.

File: api/runner/async_runner.go [hunk 5]
```diff
@@ -85,9 +85,7 @@ func deleteTask(url string, task *models.Task) error {
 // RunAsyncRunner pulls tasks off a queue and processes them
 func RunAsyncRunner(ctx context.Context, tasksrv string, tasks chan task.Request, rnr *Runner) {
 	u := tasksrvURL(tasksrv)
-
 	startAsyncRunners(ctx, u, tasks, rnr)
-	<-ctx.Done()
```
Comment (ccirello) [inline api/runner/async_runner.go]: Why did you remove this?

File: api/server/server.go [hunk 6]
```diff
@@ -177,26 +180,35 @@ func (s *Server) Start(ctx context.Context) {
...
 		Log: func(msg interface{}) {
 			logrus.Debug("supervisor: ", msg)
 		},
 	}
 
-	// By default it serves on :8080 unless a
-	// PORT environment variable was defined.
 	svr.AddFunc(func(ctx context.Context) {
-		go s.Router.Run()
+		http.Serve(listener, s.Router)
```
Comment (ccirello) [inline api/server/server.go]: How is this terminated after context cancelation? `http.Serve` is blocking and it is never going to be stopped, unless it panics or fails binding the port. And in that case, also you are not catching the error...

Comment (pedronasser) [inline api/server/server.go]: @ccirello Fixed
